File: ios/Classes/NativeAdViews/NativeAdView.swift
//
//  NativeAdView.swift
//  flutter_native_admob
//
//  Created by Dao Duy Duong on 3/8/20.
//

import UIKit
import GoogleMobileAds

enum NativeAdmobType: String {
    case banner, full
}

class NativeAdView: GADNativeAdView {
    
    let adLabelLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.textColor = .white
        label.text = "Ad"
        label.numberOfLines = 1
        return label
    }()
    
    lazy var adLabelView: UIView = {
        let view = UIView()
        view.backgroundColor = .fromHex("FFCC66")
        view.layer.cornerRadius = 3
        view.clipsToBounds = true
        view.addSubview(adLabelLbl)
        adLabelLbl.autoPinEdgesToSuperviewEdges(with: UIEdgeInsets(top: 1, left: 3, bottom: 1, right: 3))
        
        return view
    }()
    
    let adMediaView = GADMediaView()
    
    let adIconView: UIImageView = {
        let imageView = UIImageView()
        imageView.autoSetDimensions(to: CGSize(width: 40, height: 40))
        return imageView
    }()
    
    let adHeadLineLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.boldSystemFont(ofSize: 16)
        label.numberOfLines = 1
        return label
    }()
    
    let adAdvertiserLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 14)
        label.numberOfLines = 1
        return label
    }()
    
    let adRatingView = StackLayout().spacing(2)
    
    let adBodyLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.numberOfLines = 2
        return label
    }()
    
    let adPriceLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.numberOfLines = 1
        label.textAlignment = .right
        return label
    }()
    
    let adStoreLbl: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.numberOfLines = 1
        label.textAlignment = .right
        return label
    }()
    
    let callToActionBtn: UIButton = {
        let button = UIButton()
        button.setBackgroundImage(.from(color: .fromHex("#4CBE99")), for: .normal)
        button.setTitleColor(.white, for: .normal)
        button.titleLabel?.font = UIFont.systemFont(ofSize: 15)
        button.contentEdgeInsets = UIEdgeInsets(top: 0, left: 15, bottom: 0, right: 15)
        button.autoSetDimension(.height, toSize: 30)
        return button
    }()
    
    var starIcon: StarIcon {
        let icon = StarIcon()
        icon.autoSetDimensions(to: CGSize(width: 15, height: 15))
        return icon
    }
    
    var options = NativeAdmobOptions() {
        didSet { updateOptions() }
    }
    
    let type: NativeAdmobType
    
    init(frame: CGRect, type: NativeAdmobType) {
        self.type = type
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        fatalError("No support for interface builder")
    }
    
    func setNativeAd(_ nativeAd: GADNativeAd?) {
        guard let nativeAd = nativeAd else { return }
        self.nativeAd = nativeAd
        
        // Set the mediaContent on the GADMediaView to populate it with available
        // video/image asset.
        adMediaView.mediaContent = nativeAd.mediaContent
        
        // Populate the native ad view with the native ad assets.
        // The headline is guaranteed to be present in every native ad.
        adHeadLineLbl.text = nativeAd.headline
        
        adAdvertiserLbl.text = nativeAd.advertiser
        if nativeAd.advertiser.isNilOrEmpty {
            adAdvertiserLbl.isHidden = true
        }
        
        // These assets are not guaranteed to be present. Check that they are before
        // showing or hiding them.
        adBodyLbl.text = nativeAd.body
        if nativeAd.body.isNilOrEmpty {
            adBodyLbl.isHidden = true
        }
        
        callToActionBtn.setTitle(nativeAd.callToAction, for: .normal)
        if nativeAd.callToAction.isNilOrEmpty {
            callToActionBtn.isHidden = true
        }
        
        adIconView.image = nativeAd.icon?.image
        adIconView.isHidden = nativeAd.icon == nil
        
        adRatingView.arrangedSubviews.forEach { view in
            view.removeFromSuperview()
        }
        let numOfStars = Int(truncating: nativeAd.starRating ?? 0)
        adRatingView.children(Array(0..<numOfStars).map { _ in
            let icon = self.starIcon
            icon.color = options.ratingColor
            return icon
        })
        adRatingView.isHidden = nativeAd.starRating == nil
        
        adStoreLbl.text = nativeAd.store
        if nativeAd.store.isNilOrEmpty {
            adStoreLbl.isHidden = true
        }
        
        adPriceLbl.text = nativeAd.price
        if nativeAd.price.isNilOrEmpty {
            adPriceLbl.isHidden = true
        }
        
        layoutIfNeeded()
    }
}

private extension NativeAdView {
    
    func setupView() {
        self.mediaView = adMediaView
        self.headlineView = adHeadLineLbl
        self.callToActionView = callToActionBtn
        self.iconView = adIconView
        self.bodyView = adBodyLbl
        self.storeView = adStoreLbl
        self.priceView = adPriceLbl
        self.starRatingView = adRatingView
        self.advertiserView = adAdvertiserLbl
        
        switch type {
        case .full: setupFullLayout()
        case .banner: setupBannerLayout()
        }
    }
    
    func setupFullLayout() {
        adBodyLbl.numberOfLines = 2
        
        let infoLayout = StackLayout().spacing(5).children([
            adIconView,
            StackLayout().direction(.vertical).children([
                adHeadLineLbl,
                StackLayout().children([
                    adAdvertiserLbl,
                    adRatingView,
                    UIView()
                ]),
            ]),
        ])
        
        callToActionBtn.setContentHuggingPriority(.defaultHigh, for: .horizontal)
        
        let actionLayout = StackLayout()
            .spacing(5)
            .children([
                UIView(),
                adPriceLbl,
                adStoreLbl,
                callToActionBtn
            ])
        
        adBodyLbl.autoSetDimension(.height, toSize: 20, relation: .greaterThanOrEqual)
        adLabelLbl.autoSetDimension(.height, toSize: 15, relation: .greaterThanOrEqual)
        
        addSubview(adLabelView)
        adLabelView.autoPinEdge(toSuperviewEdge: .top)
        adLabelView.autoPinEdge(toSuperviewEdge: .leading)
        
        let contentView = UIView()
        addSubview(contentView)
        contentView.autoPinEdgesToSuperviewEdges(with: .zero, excludingEdge: .top)
        contentView.autoPinEdge(.top, to: .bottom, of: adLabelView, withOffset: 5)
        
        let layout = StackLayout()
            .direction(.vertical)
            .spacing(5)
            .children([
                adMediaView,
                infoLayout,
                adBodyLbl,
                actionLayout
            ])
        layout.isUserInteractionEnabled = false
        contentView.addSubview(layout)
        layout.autoAlignAxis(toSuperviewAxis: .horizontal)
        layout.autoPinEdge(toSuperviewEdge: .leading)
        layout.autoPinEdge(toSuperviewEdge: .trailing)
        layout.autoPinEdge(toSuperviewEdge: .top, withInset: 0, relation: .greaterThanOrEqual)
        layout.autoPinEdge(toSuperviewEdge: .bottom, withInset: 0, relation: .greaterThanOrEqual)
    }
    
    func setupBannerLayout() {
        adBodyLbl.numberOfLines = 2
        
        callToActionBtn.setContentHuggingPriority(.defaultHigh, for: .horizontal)
        adLabelLbl.autoSetDimension(.height, toSize: 15, relation: .greaterThanOrEqual)
        
        addSubview(adLabelView)
        adLabelView.autoPinEdge(toSuperviewEdge: .top)
        adLabelView.autoPinEdge(toSuperviewEdge: .leading)
        
        let contentView = UIView()
        addSubview(contentView)
        contentView.autoPinEdgesToSuperviewEdges(with: .zero, excludingEdge: .top)
        contentView.autoPinEdge(.top, to: .bottom, of: adLabelView, withOffset: 5)
        
        let layout = StackLayout().spacing(5).children([
            adIconView,
            StackLayout().direction(.vertical).children([
                adHeadLineLbl,
                adBodyLbl,
                StackLayout().children([
                    adAdvertiserLbl,
                    adRatingView,
                    UIView()
                ]),
            ]),
            callToActionBtn
        ])
        layout.isUserInteractionEnabled = false
        contentView.addSubview(layout)
        layout.autoAlignAxis(toSuperviewAxis: .horizontal)
        layout.autoPinEdge(toSuperviewEdge: .leading)
        layout.autoPinEdge(toSuperviewEdge: .trailing)
        layout.autoPinEdge(toSuperviewEdge: .top, withInset: 0, relation: .greaterThanOrEqual)
        layout.autoPinEdge(toSuperviewEdge: .bottom, withInset: 0, relation: .greaterThanOrEqual)
    }
    
    func updateOptions() {
        adMediaView.isHidden = !options.showMediaContent

        backgroundColor = options.backgroundColor

        adLabelLbl.textColor = options.adLabelTextStyle.color
        adLabelLbl.font = UIFont.systemFont(ofSize: options.adLabelTextStyle.fontSize)
        adLabelView.backgroundColor = options.adLabelTextStyle.backgroundColor ?? .fromHex("FFCC66")
        adLabelView.isHidden = !options.adLabelTextStyle.isVisible
        
        adHeadLineLbl.textColor = options.headlineTextStyle.color
        adHeadLineLbl.font = UIFont.systemFont(ofSize: options.headlineTextStyle.fontSize)
        adHeadLineLbl.isHidden = !options.headlineTextStyle.isVisible
        
        adAdvertiserLbl.textColor = options.advertiserTextStyle.color
        adAdvertiserLbl.font = UIFont.systemFont(ofSize: options.advertiserTextStyle.fontSize)
        adAdvertiserLbl.isHidden = !options.advertiserTextStyle.isVisible
        
        adBodyLbl.textColor = options.bodyTextStyle.color
        adBodyLbl.font = UIFont.systemFont(ofSize: options.bodyTextStyle.fontSize)
        adBodyLbl.isHidden = !options.bodyTextStyle.isVisible
        
        adStoreLbl.textColor = options.storeTextStyle.color
        adStoreLbl.font = UIFont.systemFont(ofSize: options.storeTextStyle.fontSize)
        adStoreLbl.isHidden = !options.storeTextStyle.isVisible
        
        adPriceLbl.textColor = options.priceTextStyle.color
        adPriceLbl.font = UIFont.systemFont(ofSize: options.priceTextStyle.fontSize)
        adPriceLbl.isHidden = !options.priceTextStyle.isVisible
        
        callToActionBtn.setTitleColor(options.callToActionStyle.color, for: .normal)
        callToActionBtn.titleLabel?.font = UIFont.systemFont(ofSize: options.callToActionStyle.fontSize)
        if let bgColor = options.callToActionStyle.backgroundColor {
            callToActionBtn.setBackgroundImage(.from(color: bgColor), for: .normal)
        }
        callToActionBtn.isHidden = !options.callToActionStyle.isVisible
        
        starIcon.color = options.ratingColor
    }
}

File: android/src/main/kotlin/com/nover/flutternativeadmob/NativeAdView.kt
package com.nover.flutternativeadmob

import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView

enum class NativeAdmobType {
  full, banner
}

class NativeAdView @JvmOverloads constructor(
    context: Context,
    type: NativeAdmobType,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

  var options = NativeAdmobOptions()
    set(value) {
      field = value
      updateOptions()
    }

  private val adView: NativeAdView

  private val ratingBar: RatingBar

  private var adMedia: com.google.android.gms.ads.nativead.MediaView?

  private val adHeadline: TextView
  private val adAdvertiser: TextView
  private val adBody: TextView?
  private val adPrice: TextView?
  private val adStore: TextView?
  private val adAttribution: TextView
  private val callToAction: Button

  init {
    val inflater = LayoutInflater.from(context)
    val layout = when (type) {
      NativeAdmobType.full -> R.layout.native_admob_full_view
      NativeAdmobType.banner -> R.layout.native_admob_banner_view
    }
    inflater.inflate(layout, this, true)

    setBackgroundColor(Color.TRANSPARENT)

    adView = findViewById(R.id.ad_view)

    adMedia = adView.findViewById(R.id.ad_media)

    adHeadline = adView.findViewById(R.id.ad_headline)
    adAdvertiser = adView.findViewById(R.id.ad_advertiser)
    adBody = adView.findViewById(R.id.ad_body)
    adPrice = adView.findViewById(R.id.ad_price)
    adStore = adView.findViewById(R.id.ad_store)
    adAttribution = adView.findViewById(R.id.ad_attribution)

    ratingBar = adView.findViewById(R.id.ad_stars)

    adAttribution.background = Color.parseColor("#FFCC66").toRoundedColor(3f)
    callToAction = adView.findViewById(R.id.ad_call_to_action)

    initialize()
  }

  private fun initialize() {
    // The MediaView will display a video asset if one is present in the ad, and the
    // first image asset otherwise.
    adView.mediaView = adMedia

    // Register the view used for each individual asset.
    adView.headlineView = adHeadline
    adView.bodyView = adBody
    adView.callToActionView = callToAction
    adView.iconView = adView.findViewById(R.id.ad_icon)
    adView.priceView = adPrice
    adView.starRatingView = ratingBar
    adView.storeView = adStore
    adView.advertiserView = adAdvertiser
  }

  fun setNativeAd(nativeAd: NativeAd?) {
    if (nativeAd == null) return

    // Some assets are guaranteed to be in every UnifiedNativeAd.
    adMedia?.setMediaContent(nativeAd.mediaContent)
    adMedia?.setImageScaleType(ImageView.ScaleType.FIT_CENTER)

    adHeadline.text = nativeAd.headline
    adBody?.text = nativeAd.body
    (adView.callToActionView as Button).text = nativeAd.callToAction

    // These assets aren't guaranteed to be in every UnifiedNativeAd, so it's important to
    // check before trying to display them.
    val icon = nativeAd.icon

    if (icon == null) {
      adView.iconView?.visibility = View.GONE
    } else {
      (adView.iconView as ImageView).setImageDrawable(icon.drawable)
      adView.iconView?.visibility = View.VISIBLE
    }

    if (nativeAd.price == null) {
      adPrice?.visibility = View.INVISIBLE
    } else {
      adPrice?.visibility = View.VISIBLE
      adPrice?.text = nativeAd.price
    }

    if (nativeAd.store == null) {
      adStore?.visibility = View.INVISIBLE
    } else {
      adStore?.text = nativeAd.store
    }

    if (nativeAd.starRating == null) {
      adView.starRatingView?.visibility = View.INVISIBLE
    } else {
      (adView.starRatingView as RatingBar).rating = nativeAd.starRating!!.toFloat()
      adView.starRatingView?.visibility = View.VISIBLE
    }

    if (nativeAd.advertiser == null) {
      adAdvertiser.visibility = View.INVISIBLE
    } else {
      adAdvertiser.visibility = View.VISIBLE
      adAdvertiser.text = nativeAd.advertiser
    }

    // Assign native ad object to the native view.
    adView.setNativeAd(nativeAd)
  }

  private fun updateOptions() {
    adView.setBackgroundColor(options.backgroundColor)

    adMedia?.visibility = if (options.showMediaContent) View.VISIBLE else View.GONE

    ratingBar.progressDrawable
        .setColorFilter(options.ratingColor, PorterDuff.Mode.SRC_ATOP)

    options.adLabelTextStyle.backgroundColor?.let {
      adAttribution.background = it.toRoundedColor(3f)
    }
    adAttribution.textSize = options.adLabelTextStyle.fontSize
    adAttribution.setTextColor(options.adLabelTextStyle.color)
    adAdvertiser.visibility = options.adLabelTextStyle.visibility
    options.adLabelTextStyle.androidTypeface?.let { t ->
      adAttribution.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }

    adHeadline.setTextColor(options.headlineTextStyle.color)
    adHeadline.textSize = options.headlineTextStyle.fontSize
    adHeadline.visibility = options.headlineTextStyle.visibility
    options.headlineTextStyle.androidTypeface?.let { t ->
      adHeadline.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }

    adAdvertiser.setTextColor(options.advertiserTextStyle.color)
    adAdvertiser.textSize = options.advertiserTextStyle.fontSize
    adAdvertiser.visibility = options.advertiserTextStyle.visibility
    options.advertiserTextStyle.androidTypeface?.let { t ->
      adAdvertiser.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }


    adBody?.setTextColor(options.bodyTextStyle.color)
    adBody?.textSize = options.bodyTextStyle.fontSize
    adBody?.visibility = options.bodyTextStyle.visibility
    options.bodyTextStyle.androidTypeface?.let { t ->
      adBody?.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }

    adStore?.setTextColor(options.storeTextStyle.color)
    adStore?.textSize = options.storeTextStyle.fontSize
    adStore?.visibility = options.storeTextStyle.visibility
    options.storeTextStyle.androidTypeface?.let { t ->
      adStore?.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }

    adPrice?.setTextColor(options.priceTextStyle.color)
    adPrice?.textSize = options.priceTextStyle.fontSize
    adPrice?.visibility = options.priceTextStyle.visibility
    options.priceTextStyle.androidTypeface?.let { t ->
      adPrice?.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }

    callToAction.setTextColor(options.callToActionStyle.color)
    callToAction.textSize = options.callToActionStyle.fontSize
    options.callToActionStyle.backgroundColor?.let {
      callToAction.setBackgroundColor(it)
    }
    options.callToActionStyle.androidTypeface?.let { t ->
      callToAction.typeface = Typeface.createFromAsset(
              context.assets, t
      )
    }
  }
}